CONTRIBUTING.md

```diff
@@ -29,3 +29,142 @@ Recommendations and requirements for how to best contribute to Parquet. We striv
+
+3. Forward incompatible. A file written under a newer version of the format with
+   the feature enabled cannot be read under an older version of the format (e.g.
+   adding and using a new compression algorithm). It is expected any feature in
+   this category will provide a signal to older readers, so they can
+   unambiguously determine that they cannot properly read the file (e.g. via
+   adding a new value to an existing enum).
+
+New features are intended to be widely beneficial to users of Parquet, and
+therefore it is hoped third-party implementations will adopt them quickly after
+they are introduced. It is assumed that writing new parts of the format, and
+especially forward incompatible features, will be configured with a feature flag
+defaulted to "off", and at some future point the feature is turned on by default
+(reading of the new feature will typically be enabled without configuration or
+defaulted to on). Some amount of lead time is desirable to ensure a critical
+mass of Parquet implementations support a feature to avoid compatibility issues
+across the ecosystem.  Therefore, the Parquet PMC gives the following
+recommendations for managing features:
+
+1. Backward compatibility is the concern of implementations but given the
+   ubiquity of Parquet and the length of time it has been used, libraries should
+   support reading older versions of the format to the greatest extent possible.
+
+2. Forward compatible features/changes may be enabled and used by default in
+   implementations once the parquet-format containing those changes has been
+   formally released.  For features that may pose a significant performance
+   regression to older format readers, libaries should consider delaying default
+   enablement until 1 year after the release of the parquet-java implementation
+   that contains the feature implementation.
+
+3. Forward incompatible features/changes should not be turned on by default
+   until 2 years after the parquet-java implementation containing the feature is
+   released. It is recommended that changing the default value for a forward
+   incompatible feature flag should be clearly advertised to consumers (e.g. via
+   a major version release if using Semantic Versioning, or highlighed in
+   release notes).
+
+For forward compatible changes which have a high chance of performance
+regression for older readers and forward incompatible changes, implementations
+should clearly document the compatibility issues. Additionally, while it is up
+to maintainers of individual implementations to make the best decision to serve
+their ecosystem, they are encouraged to start enabling features by default along
+the same timelines as `parquet-java`.  Parquet-java will wait to enable features
+by default until the most conservative timelines outlined above have been
+exceeded.
+
+For features released prior to October 2024, target dates for each of these
```

Comment:
I don't quite understand this. Do we expect a parquet-java 2.0 release around October 2024? Doesn't any new feature should follow this guidance once published?

Reply:
My intent was to try for Parquet-java 2.0 in October.   There are a lot of old features from the spec perspective that based on guidance would be ok to use (e.g. lz4-raw, byte stream split, data page v2, v2 encodings) that we need to decide on recommended dates (as a sample suggestion I would aim for 1 year for features we think we actually want to turn on)

Reply:
Is it a good timing to do Parquet-java 2.0 release? As per the discussion from https://lists.apache.org/thread/kttwbl5l7opz6nwb5bck2gghc2y3td0o, we intend to break the API compatibility by removing deprecated APIs (and remove Java 8 support) in the 2.0 release. If the intention is to set a target date for existing features without breaking anything, perhaps we should release 1.15.0 instead?

cc @gszadovszky @Fokko @julienledem

Reply:
IMO I think API compatibility is a separate concern and it is fine to set future dates and change API compatibility.  I can rephrase this section to mention some other criteria for initial target dates if there is an alternative proposal?

Reply:
I would be up for a Parquet 2.0 release in October and am eager to help! Do we also want a [1.15](https://github.com/apache/parquet-java/milestone/18)? 

I'm trying to downstream 1.14.1 into various projects (https://github.com/apache/spark/pull/46447, https://github.com/apache/iceberg/pull/10209), but the new Jackson version is giving us some troubles. It feels to me that the different projects need some time to update. My argument is that releasing 1.15 right now would not give too many new features to the users, and I think it makes sense to jump to 2.0.0.

Since there are some delays mentioned earlier in this document:

- The API that I suggested removing was marked as deprecated almost six years go: https://github.com/apache/parquet-java/pull/535
- We're still compiling against Hadoop 2.7.3 which was released in August 2016.
- I think the Java8 deprecation might be a bit harder and might require more discussion, but there it feels like all the projects are looking at each other :) Spark/Avro are Java 17+ right now.

For my understanding, the things you mentioned (e.g. lz4-raw, byte stream split, data page v2, v2 encodings), they are already shipped with the last release, so the discussion is when to make them the default?

Reply:
> For my understanding, the things you mentioned (e.g. lz4-raw, byte stream split, data page v2, v2 encodings), they are already shipped with the last release, so the discussion is when to make them the default?

Yes, it is mostly to pick a time-frame where we feel comfortable to make them a default option for choosing when writing new parquet files.  I think at least for encodings we need to convince ourselves which ones are good choices based on benchmarks to be turned on by default.  For data page v2, I think dremio tried enabling this by default and then reverted it (IIRC performance on write, so there might be more work here).

Reply:
Am I correct in understanding that We intend to:
 - increment the major release of parquet-format when it has the definition of the new (forward incompatible) feature in it.
 - implement the above new feature in parquet-java under a minor release in the previous major but turning it off by default
 - increment the major release of parquet-java when we turn on the new encodings by default?

I think this is not unsound but it might be confusing. We should then expect to have several major releases coming: one for the new footer and one for each new encoding.

Reply:
@julienledem no I think there is a misunderstanding.

> increment the major release of parquet-format when it has the definition of the new (forward incompatible) feature in it.

Per above, I don't think we expect a major version change for parquet-format in the near future.

> implement the above new feature in parquet-java under a minor release in the previous major but turning it off by default

yes, this is correct.  this provides read support.

> increment the major release of parquet-java when we turn on the new encodings by default?

Yes. 


> I think this is not unsound but it might be confusing. We should then expect to have several major releases coming: one for the new footer and one for each new encoding.

I think these should be grouped.  Per https://github.com/apache/parquet-site/pull/61 the proposal for parquet-java is to have at most one major release per year.  Early adopters can use features once a minor release containing them is out as long as they feel comfortable with understanding the scope of impact.

Reply:
Sorry, I'm not following which part of the conversation "per above" refers to here.
Could you point to it and explain what you mean?

I was assuming for example that once we define a new footer in a new format we would increment the format spec. Would that be wrong?

Reply:
In the releases section:

```
The Parquet PMC aims to do releases of the format package only as needed when
new features are introduced. If multiple new features are being proposed
simultaneously some features might be consolidated into the same release.
Guidance is provided below on when implementations should enable features added
to the specification.  Due to confusion in the past over Parquet versioning it
is not expected that there will be a 3.x release of the specification in the
foreseeable future.
```

Reply:
OK, that sounds good to me.